overlord/configstate/configcore/services.go

```diff
@@ -231,6 +239,144 @@ func handleServiceDisableConfiguration(_ sysconfig.Device, tr config.ConfGetter,
 			}
 		}
 	}
+	// configure ssh ports
+	if err := handleServiceConfigSSHListen(dev, tr, opts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func parseOneSSHListenAddr(oneAddr string) (addrs []string, err error) {
+	// 1. check if it's something like "host:port", "[host]:port" etc
+	//    (this will return an error if there is no
+	host, portStr, err := net.SplitHostPort(oneAddr)
+	if err != nil {
+		// for any error assume there is no port and continue
+		host = oneAddr
+	}
+	// 2. valid port (if needed)
+	if portStr != "" {
+		port, err := strconv.Atoi(portStr)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse port number: %v", err)
+		}
+		if port > 65535 || port < 1 {
+			return nil, fmt.Errorf("cannot use port %v: must be in the range 1-65535", port)
+		}
+	}
+	// 3. validate host
+	if host != "" {
+		if net.ParseIP(host) == nil && validateHostname(host) != nil {
+			return nil, fmt.Errorf("cannot use %q as hostname", host)
+		}
+	}
+
+	// at this point the oneAddr is validated but openssh will
+	// error when no host is given, so workaround here
+	if host == "" && portStr != "" {
+		return []string{
+			fmt.Sprintf("0.0.0.0:%v", portStr),
+			fmt.Sprintf("[::]:%v", portStr),
+		}, nil
+	}
+
+	// no special handling needed and a valid listen address
+	return []string{oneAddr}, nil
+}
+
+func parseSSHListenCfg(cfgStr string) ([]string, error) {
+	var listenAddrs []string
+	for _, hostAndPort := range strings.Split(cfgStr, ",") {
+		addrs, err := parseOneSSHListenAddr(hostAndPort)
+		if err != nil {
+			return nil, err
+		}
+		listenAddrs = append(listenAddrs, addrs...)
+	}
+	return listenAddrs, nil
+}
+
+func validateServiceConfiguration(tr config.ConfGetter) error {
+	// validate the ssh listen setting
+	output, err := coreCfg(tr, sshListenOpt)
+	if err != nil {
+		return err
+	}
+	if output == "" {
+		return nil
+	}
+	if _, err := parseSSHListenCfg(output); err != nil {
+		return err
```

Comment:
should this be prefixed with something?

Reply:
another option is to prefix with "ssh" port and hostname etc in the errors in parseOneSSHListenAddr?

Reply:
Thanks, I added an error prefix now.